py/makemoduledefs.py

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/env python
+
+# This pre-processor parses provided objects' c files for 
+# `MP_REGISTER_MODULE(module_name, obj_module, enabled_define)`
+# These are used to generate a header with the required entries for 
+# `mp_rom_map_elem_t mp_builtin_module_table[]` in py/objmodule.c
+
+from __future__ import print_function
+
+import re
+import os
+import argparse
+
+
+def find_c_file(vpath, obj_file):
+    """
+    Search the provided vpaths for the c file that matches the provided object_file
+    :param List[str} vpath: List of base paths, similar to gcc vpath
+    :param obj_file: object file to find the matching c file for
+    :return: path to c file or None
+    """
+    c_file = None
+    relative_c_file = os.path.splitext(obj_file)[0] + ".c"
+    relative_c_file = relative_c_file.lstrip('/\\')
+    for p in vpath:
+        possible_c_file = os.path.join(p, relative_c_file)
+        if os.path.exists(possible_c_file):
+            c_file = possible_c_file
+            break
+
+    return c_file
+
+
+def find_module_registrations(c_file):
+    """
+    Find any MP_REGISTER_MODULE definitions in the provided c file
+    :param c_file: path to c file to check
+    :return: List((module_name, obj_module, enabled_define))
+    """
+    pattern = re.compile(
+        r"[\n;]\s*MP_REGISTER_MODULE\((.*?),\s*(.*?),\s*(.*?)\);",
+        flags=re.DOTALL
+    )
+
+    if c_file is None:
+        # No c file to match the object file, skip
+        return []
+
+    with open(c_file) as c_file_handle:
+        c_text = c_file_handle.read()
```

Comment:
why not just return list(re.findall(pattern, c_file_handle.read())) here?
Minor nitpick: open() doesn't really return a 'handle' like there's handles in C, but it's called a 'file object'

Reply:
Splitting it up was useful while debugging to get the pattern right, less important now!